src/NATS.Client.Core/Commands/CommandWriter.cs

```diff
@@ -262,8 +297,92 @@ public async ValueTask UnsubscribeAsync(int sid, int? maxMsgs, CancellationToken
     // only used for internal testing
     internal bool TestStallFlush() => _channelLock.Writer.TryWrite(1);
 
-    [MethodImpl(MethodImplOptions.NoInlining)]
-    private static void ThrowOnDisconnected() => throw new NatsException("Connection hasn't been established yet.");
+    private static async Task ReaderLoopAsync(ILogger<CommandWriter> logger, ISocketConnection connection, PipeReader pipeReader, Channel<int> channelSize, CancellationToken cancellationToken)
+    {
+        try
+        {
+            while (true)
+            {
+                var result = await pipeReader.ReadAsync(cancellationToken).ConfigureAwait(false);
+
+                if (result.IsCanceled)
+                {
+                    break;
+                }
+
+                var buffer = result.Buffer;
```

Comment:
need to slice the Examined Offset off the front of the buffer here.  the Examined Offset are the bytes that were sent last iteration, but not marked as consumed

```
var buffer = result.Buffer.Slice(examinedOffset)
```